atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java

```diff
@@ -44,18 +48,36 @@ default LockWatchVersion lastVersion() {
 
     @Value.Derived
     default UUID leader() {
-        return timestampMapping().values().stream()
-                .map(LockWatchVersion::id)
-                .distinct()
-                .collect(MoreCollectors.onlyElement());
+        // explicitly avoiding stream and distinct collection copy
+        return distinctElement(Collections2.transform(timestampMapping().values(), LockWatchVersion::id));
     }
 
     @Value.Check
     default void nonEmptyMapping() {
-        Preconditions.checkArgument(!timestampMapping().isEmpty(), "Cannot process an empty timestamp map");
+        checkNotEmpty(timestampMapping());
+    }
+
+    private Map<Long, LockWatchVersion> checkNotEmpty(Map<Long, LockWatchVersion> timestampMapping) {
+        Preconditions.checkArgument(!timestampMapping.isEmpty(), "Cannot process an empty timestamp map");
+        return timestampMapping;
     }
 
     static ImmutableTimestampMapping.Builder builder() {
         return ImmutableTimestampMapping.builder();
     }
+
+    @VisibleForTesting
+    static <T> T distinctElement(Iterable<? extends T> iterable) {
+        Iterator<? extends T> iterator = iterable.iterator();
+        if (!iterator.hasNext()) {
+            throw new SafeIllegalStateException("No elements");
```

Comment:
Noting for posterity, but I don't actually think it matters - we're changing from MoreCollectors NoSuchElementException and IllegalArgumentException to IllegalStateException

Reply:
updated to maintain exception back compat (though throwing SafeIllegalArgumentException with mismatched IDs)